File: ollama/codigo-final/fasta/llama3.1-8b/cot/temp_0/fasta.py
import random

# Define LCG parameters
IM = 139968
IA = 3877
IC = 29573
Seed = 42

def lcg(seed):
    """Linear Congruential Generator"""
    global IM, IA, IC
    seed = (seed * IA + IC) % IM
    return seed

# Define ALU sequence
alu_sequence = "GGCCGGGCGCGGTGGCTCACGCCTGTAATCCCAGCACTTTGGGAGGCCGAGGCGGGCGGA" \
               "TCACCTGAGGTCAGGAGTTCGAGACCAGCCTGGCCAACATGGTGAAACCCCGTCTCTACTAAAAAATACAAAAATTAGCCGGGCGTGGTGGCGCGCGCCTGTAATCCCAGCTACTCGGGAGGCTGAGGCAGGAGAATCGCTTGAACCCGGGAGGCGGAGGTTGCAGTGAGCCGAGATCGCGCCACTGCACTCCAGCCTGGGCGACAGAGCGAGACTCCGTCTCAAAAA"

# Define IUB probabilities
iub_probabilities = {
    'a': 0.27,
    'c': 0.12,
    'g': 0.12,
    't': 0.27,
    'B': 0.02,
    'D': 0.02,
    'H': 0.02,
    'K': 0.02,
    'M': 0.02,
    'N': 0.02,
    'R': 0.02,
    'S': 0.02,
    'V': 0.02,
    'W': 0.02
}

# Define Homo sapiens probabilities
homo_sapiens_probabilities = {
    'a': 0.3029549426680,
    'c': 0.1979883004921,
    'g': 0.1975473066391,
    't': 0.3015094502008
}

def generate_sequence(probabilities, length):
    """Generate a random sequence based on given probabilities"""
    sequence = ''
    for _ in range(length):
        r = lcg(Seed)
        Seed = r
        total_prob = sum(probabilities.values())
        rand_num = (r % 100) / 100
        cumulative_prob = 0
        for base, prob in probabilities.items():
            cumulative_prob += prob / total_prob
            if rand_num < cumulative_prob:
                sequence += base
                break
    return sequence

def generate_fasta(n):
    """Generate three DNA sequences based on specific rules and probabilities"""
    Seed = 42
    alu_sequence = alu_sequence * (n*2)
    iub_sequence = generate_sequence(iub_probabilities, n*3)
    homo_sapiens_sequence = generate_sequence(homo_sapiens_probabilities, n*5)

    with open('output.txt', 'w') as f:
        f.write("ONE Homo sapiens alu\n")
        for i in range(0, len(alu_sequence), 60):
            f.write(alu_sequence[i:i+60] + "\n")
        f.write("\nTWO IUB ambiguity codes\n")
        for i in range(0, len(iub_sequence), 60):
            f.write(iub_sequence[i:i+60] + "\n")
        f.write("\nTHREE Homo sapiens frequency\n")
        for i in range(0, len(homo_sapiens_sequence), 60):
            f.write(homo_sapiens_sequence[i:i+60] + "\n")

generate_fasta(1000)

File: ollama/codigo-final/fasta/llama3.1-8b/cot/temp_0/fasta.cpp
#include <iostream>
#include <string>
#include <vector>

// Function to generate a random number using LCG
int lcg(int seed) {
    return (seed * 1103515245 + 12345) % 139968;
}

// Function to generate ALU sequence
std::string alu(int n) {
    std::string seq = "GGCCGGGCGCGGTGGCTCACGCCTGTAATCCCAGCACTTTGGGAGGCCGAGGCGGGCGGA TCACCTGAGGTCAGGAGTTCGAGACCAGCCTGGCCAACATGGTGAAACCCCGTCTCTACTAAAAA ATACAAAAATTAGCCGGGCGTGGTGGCGCGCGCCTGTAATCCCAGCTACTCGGGAGGCTGAGGC AGGAGAATCGCTTGAACCCGGGAGGCGGAGGTTGCAGTGAGCCGAGATCGCGCCACTGCACTCCAG CTTGGGCGACAGAGCGAGACTCCGTCTCAAAAA";
    std::string result;
    for (int i = 0; i < n * 2; ++i) {
        int r = lcg(42);
        if (r % 4 == 0) result += 'A';
        else if (r % 4 == 1) result += 'C';
        else if (r % 4 == 2) result += 'G';
        else result += 'T';
    }
    return seq + "\n" + result;
}

// Function to generate IUB sequence
std::string iub(int n) {
    std::vector<std::pair<char, double>> bases = {{'A', 0.27}, {'C', 0.12}, {'G', 0.12}, {'T', 0.27},
        {'B', 0.02}, {'D', 0.02}, {'H', 0.02}, {'K', 0.02}, {'M', 0.02}, {'N', 0.02},
        {'R', 0.02}, {'S', 0.02}, {'V', 0.02}, {'W', 0.02}};
    std::string result;
    for (int i = 0; i < n * 3; ++i) {
        double r = static_cast<double>(lcg(42)) / 139968;
        char base;
        for (const auto& b : bases) {
            if (r <= b.second) {
                base = b.first;
                break;
            }
            r -= b.second;
        }
        result += base;
    }
    return "TWO IUB ambiguity codes\n" + result;
}

// Function to generate Homo sapiens sequence
std::string homo(int n) {
    std::vector<std::pair<char, double>> bases = {{'A', 0.3029549426680}, {'C', 0.1979883004921},
        {'G', 0.1975473066391}, {'T', 0.3015094502008}};
    std::string result;
    for (int i = 0; i < n * 5; ++i) {
        double r = static_cast<double>(lcg(42)) / 139968;
        char base;
        for (const auto& b : bases) {
            if (r <= b.second) {
                base = b.first;
                break;
            }
            r -= b.second;
        }
        result += base;
    }
    return "THREE Homo sapiens frequency\n" + result;
}

int main(int argc, char* argv[]) {
    int n = std::stoi(argv[1]);
    std::cout << alu(n) << iub(n) << homo(n);
    return 0;
}
